crates/walrus-service/src/node/blob_event_processor.rs

```diff
@@ -316,12 +391,58 @@ impl BlobEventProcessor {
             )
             .inc();
 
+            // Increment the counter and create a guard that will decrement it when dropped
+            let current_pending_event_count = self.background_per_processor_pending_event_count
+                [processor_index]
+                .fetch_add(1, Ordering::SeqCst)
+                + 1;
+            walrus_utils::with_label!(
+                self.node
+                    .metrics
+                    .pending_processing_blob_event_in_background_processors,
+                &processor_index.to_string()
+            )
+            .set(i64::try_from(current_pending_event_count).unwrap_or(i64::MAX));
+
+            // Create the guard that will automatically decrement the counter when dropped
+            let guard = PendingEventGuard::new(
+                self.background_per_processor_pending_event_count[processor_index].clone(),
+                processor_index,
+                self.node.metrics.clone(),
+            );
+
+            // Send the wrapped event with the guard
+            let tracked_event = TrackedEvent {
+                event_handle,
+                blob_event,
+                _guard: guard,
+            };
+
             self.background_processor_senders[processor_index]
-                .send((event_handle, blob_event))
+                .send(tracked_event)
                 .map_err(|e| {
                     anyhow::anyhow!("failed to send event to background processor: {}", e)
                 })?;
         }
         Ok(())
     }
+
+    /// Waits for all events to be processed in the background processors.
+    pub async fn wait_for_all_events_to_be_processed(&self) {
+        if self.background_processor_senders.is_empty() {
+            // When there are no background workers, we use a sequential processor to process events
+            // sequentially in the same thread, and no events are processed in the background.
+            // Therefore, we don't need to wait, and can return immediately.
+            return;
+        }
+
+        // Check if any of the background processors still have pending events.
+        while self
+            .background_per_processor_pending_event_count
+            .iter()
+            .any(|c| c.load(Ordering::SeqCst) > 0)
+        {
+            tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
Suggestion: Decrease this sleep duration to something like 5-10ms. The check is pretty fast, so it doesn't hurt to check more often. Also, this could be an explicit constant or a configuration parameter.

Reply:
Or 1ms. This check is going to run in probably less than 1 microsecond. How long are these background processes running generally?